### PR TITLE
Implemented support for SMT transfers

### PR DESCRIPTION
**Pull request by Kiwonik**
Modified `adjust_balance`, `adjust_reward_balance` and `get_balance`. Refactored `savings` to `rewards`. Added simple test.  #1682

### CODE REVIEW

**Comment by youkaicountry** [comment]
Looks like there are merge conflicts though.

**Comment by Kiwonik** [comment]
@youkaicountry I rebased the branch on develop's HEAD. Ready to merge again.

**Comment by Kiwonik** [comment]
@vogel76 The code you commented is used by `adjust_balance` (modify) method, not by `get_balance` (check).

File: libraries/chain/database.cpp [hunk 1]
```diff
@@ -3404,15 +3405,58 @@ void database::clear_expired_delegations()
       itr = delegations_by_exp.begin();
    }
 }
+#ifdef STEEM_ENABLE_SMT
+template< typename smt_balance_object_type >
+void database::adjust_smt_balance( const account_object& a, const asset& delta, bool check_balance )
```
Comment (mvandeberg) [inline libraries/chain/database.cpp]: Only the account name is used in this method, so let's just pass the account name.

File: libraries/chain/database.cpp [hunk 2]
```diff
@@ -3404,15 +3405,58 @@ void database::clear_expired_delegations()
...
+   {
+      modify( *bo, [&]( smt_balance_object_type& smt_balance )
+      {
+         smt_balance.balance += delta;
+      } );
+   }
 
+   if( check_balance )
+   {
+      FC_ASSERT( bo->balance.amount.value >= 0, "Insufficient SMT ${smt} funds", ("smt", delta.symbol) );
```
Comment (mvandeberg) [inline libraries/chain/database.cpp]: This check is incorrect and should always be done, regardless of the existence of the account.

What if the delta is negative? Then a balance object with a negative balance will be created.

File: libraries/chain/database.cpp [hunk 3]
```diff
@@ -3404,15 +3405,58 @@ void database::clear_expired_delegations()
...
+   }
+}
```
Comment (mvandeberg) [inline libraries/chain/database.cpp]: This method should probably automatically remove balance objects if the resulting balance is 0.

File: libraries/chain/database.cpp (hunk 2 above)
Comment (mvandeberg) [reply libraries/chain/database.cpp]: I see it is passed from `adjust_balance`. Because SMTs are new, we don't need to gate it with a hardfork check.

**Comment by mvandeberg** [comment]
This is challenging to review because `develop` was merged into the feature branch rather than the feature branch being rebased on `develop`. All of the commits that were added to `develop` since the base of the branch are now showing in the diff.

**Comment by Kiwonik** [comment]
Indeed they are, even though I **did** do `git rebase develop` with the feature branch and not merging :(
I'll make another cherry-picked branch and add requested changes to it.

**Comment by Kiwonik** [comment]
[PR 1842](https://github.com/steemit/steem/pull/1842) replaces this one.

**Comment by youkaicountry** [comment]
Closing in favor of #1842 replacing this.

**Comment by youkaicountry** [comment]
@mvandeberg @theoreticalbts What is the policy on closed PR branches. Should I delete it as I close the PR?

**Comment by mvandeberg** [comment]
Only delete it if you know we don't need the branch. I periodically clean up old branches, so if one slips through, it will get cleaned up eventually.

**Pull request closed without merging**